### PR TITLE
Core/adding dofs interface

### PR DESCRIPTION
**Pull request by RiccardoRossi**
this PR adds the interface of dofs and of DofsArray to the model_part.

this is another step to allow working with Kratos matrices from python.

The next step will be to revive an old discussion about giving access to mass, stiffness and damping matrices so that they can be manipulated from python

### CODE REVIEW

**Comment by RiccardoRossi** [comment]
just to add that this will have to wait for #6096 before being merged (mostly because i wrongly derived one PR from the other, they are actually quite independent)

File: kratos/python/add_strategies_to_python.cpp [hunk 1]
```diff
@@ -12,9 +12,8 @@
 
 // System includes
 
-
 // External includes
-#ifdef KRATOS_USE_AMATRIX   // This macro definition is for the migration period and to be removed afterward please do not use it
+#ifdef KRATOS_USE_AMATRIX                 // This macro definition is for the migration period and to be removed afterward please do not use it
```
Comment (loumalouomega) [inline kratos/python/add_strategies_to_python.cpp]: The file changes are confusing. What happened here?

File: kratos/python/add_node_to_python.cpp [hunk 2]
```diff
@@ -141,6 +141,16 @@ void  AddNodeToPython(pybind11::module& m)
     ;
 
     py::class_<Dof<double>>(m,"Dof")
+    .def("GetVariable", &Dof<double>::GetVariable, py::return_value_policy::reference_internal)
```
Comment (loumalouomega) [inline kratos/python/add_node_to_python.cpp]: Finally!!!, we should expose the construction of MPC to python now that dof are fully available in python. Will be very practical

**Review by loumalouomega — COMMENTED** [review]
Some comments

**Comment by armingeiser** [comment]
This does partially solve #5897.

File: kratos/python/add_node_to_python.cpp (hunk 2 above)
Comment (RiccardoRossi) [reply kratos/python/add_node_to_python.cpp]: ok, but only after this PR is merged

File: kratos/python/add_strategies_to_python.cpp [hunk 3]
```diff
@@ -371,8 +371,40 @@ namespace Kratos
             typedef BuilderAndSolver< SparseSpaceType, LocalSpaceType, LinearSolverType > BuilderAndSolverType;
 
 
-            py::class_< BuilderAndSolverType::DofsArrayType, BuilderAndSolverType::DofsArrayType::Pointer>(m,"DofsArrayType")
-                .def(py::init<>());
+            py::class_<BuilderAndSolverType::DofsArrayType, BuilderAndSolverType::DofsArrayType::Pointer>(m, "DofsArrayType")
```
Comment (RiccardoRossi) [inline kratos/python/add_strategies_to_python.cpp]: for whatever reason it formatted the file, so there were clashes all over

the only changes needed are these ones

**Review by RiccardoRossi — COMMENTED** [review]
corrected to remove formatting changes

File: kratos/python/add_strategies_to_python.cpp (hunk 1 above)
Comment (RiccardoRossi) [reply kratos/python/add_strategies_to_python.cpp]: wrong formatting, removed

**Comment by RiccardoRossi** [comment]
@armingeiser i remember that some of you guys had an issue about exporting M, K, D matrices to python.

that was closed long ago, since no one stepped up for the implementation. do you happen to remember were the discussion was? i would like to resume it as a follow up of this

**Comment by armingeiser** [comment]
> @armingeiser i remember that some of you guys had an issue about exporting M, K, D matrices to python.
> 
> that was closed long ago, since no one stepped up for the implementation. do you happen to remember were the discussion was? i would like to resume it as a follow up of this

I don't remember to be honest, i can have a more detailed look tomorrow. Maybe it was related to the testing of the eigensystem solver of the EigenSolversApplication. I think that was solved finally by @oberbichler implementing the matrixmarket interface to the Kratos sparse matrices.

**Comment by oberbichler** [comment]
@armingeiser we found a solution to test them without implementing the MatrixMarket-IO

File: kratos/python/add_strategies_to_python.cpp [hunk 4]
```diff
@@ -371,8 +371,40 @@ namespace Kratos
...
+                .def(py::init<>())
+                .def("__len__", [](BuilderAndSolverType::DofsArrayType &self) { return self.size(); })
+                .def("__iter__", [](BuilderAndSolverType::DofsArrayType &self) { return py::make_iterator(self.begin(), self.end()); }, py::keep_alive<0, 1>())
+                .def("append", [](BuilderAndSolverType::DofsArrayType &self,
+                                Dof<double> &value) { self.push_back(&value); })
+                .def("GetValues", [](BuilderAndSolverType::DofsArrayType &self) {
+                    Vector values(self.size());
+                    int counter = 0;
+                    for (auto &dof : self)
```
Comment (rubenzorrilla) [inline kratos/python/add_strategies_to_python.cpp]:
```suggestion
                    for (auto &r_dof : self)
```

File: kratos/python/add_strategies_to_python.cpp [hunk 5]
```diff
@@ -371,8 +371,40 @@ namespace Kratos
...
+                    {
+                        values[counter++] = dof.GetSolutionStepValue();
```
Comment (rubenzorrilla) [inline kratos/python/add_strategies_to_python.cpp]:
```suggestion
                        values[counter++] = r_dof.GetSolutionStepValue();
```

File: kratos/python/add_strategies_to_python.cpp [hunk 6]
```diff
@@ -371,8 +371,40 @@ namespace Kratos
...
+                    }
+                    return values;
+                })
+                .def("GetEquationIds", [](BuilderAndSolverType::DofsArrayType &self) {
+                    std::vector<std::size_t> values(self.size());
+                    int counter = 0;
+                    for (auto &dof : self)
```
Comment (rubenzorrilla) [inline kratos/python/add_strategies_to_python.cpp]:
```suggestion
                    for (auto &r_dof : self)
```

File: kratos/python/add_strategies_to_python.cpp [hunk 7]
```diff
@@ -371,8 +371,40 @@ namespace Kratos
...
+                    {
+                        values[counter++] = dof.EquationId();
```
Comment (rubenzorrilla) [inline kratos/python/add_strategies_to_python.cpp]:
```suggestion
                        values[counter++] = r_dof.EquationId();
```

File: kratos/python/add_strategies_to_python.cpp [hunk 8]
```diff
@@ -371,8 +371,40 @@ namespace Kratos
...
+                    }
+                    return values;
+                })
+                .def("SetValues", [](BuilderAndSolverType::DofsArrayType &self,
+                                    const Vector& values
+                                    ) {
+                    int counter = 0;
+                    for (auto &dof : self)
```
Comment (rubenzorrilla) [inline kratos/python/add_strategies_to_python.cpp]:
```suggestion
                    for (auto &r_dof : self)
```

File: kratos/python/add_strategies_to_python.cpp [hunk 9]
```diff
@@ -371,8 +371,40 @@ namespace Kratos
...
+                    {
+                        dof.GetSolutionStepValue() = values[counter++];
```
Comment (rubenzorrilla) [inline kratos/python/add_strategies_to_python.cpp]:
```suggestion
                        r_dof.GetSolutionStepValue() = values[counter++];
```

File: kratos/python/add_strategies_to_python.cpp [hunk 10]
```diff
@@ -371,8 +371,40 @@ namespace Kratos
...
+                    std::vector<std::size_t> values(self.size());
+                    int counter = 0;
+                    for (auto &dof : self)
+                    {
+                        values[counter++] = dof.EquationId();
+                    }
+                    return values;
+                })
+                .def("SetValues", [](BuilderAndSolverType::DofsArrayType &self,
+                                    const Vector& values
```
Comment (rubenzorrilla) [inline kratos/python/add_strategies_to_python.cpp]:
```suggestion
                                    const Vector& values)
```

File: kratos/python/add_strategies_to_python.cpp [hunk 11]
```diff
@@ -371,8 +371,40 @@ namespace Kratos
...
+                                    ) {
```
Comment (rubenzorrilla) [inline kratos/python/add_strategies_to_python.cpp]:
```suggestion
                                {
```

File: kratos/python/add_node_to_python.cpp [hunk 12]
```diff
@@ -141,6 +141,16 @@ void  AddNodeToPython(pybind11::module& m)
     ;
 
     py::class_<Dof<double>>(m,"Dof")
```
Comment (rubenzorrilla) [inline kratos/python/add_node_to_python.cpp]: At this point I think it makes sense to create an add_dof_to_python.cpp

**Review by rubenzorrilla — COMMENTED** [review]
Some minor comments.

File: kratos/python/add_node_to_python.cpp (hunk 12 above)
Comment (loumalouomega) [reply kratos/python/add_node_to_python.cpp]: +1 to @rubenzorrilla I was thinking the same. I don't like when the python definition are mixed in an inconsistent way and I expend time to find the file where is defined

Comment (philbucher) [reply kratos/python/add_node_to_python.cpp]: +1, it is often quite tricky to find the py-exposure things

Comment (armingeiser) [reply kratos/python/add_node_to_python.cpp]: +1

Comment (RiccardoRossi) [reply kratos/python/add_node_to_python.cpp]: done

**Review by rubenzorrilla — APPROVED** [review]
No more concerns from my side

**Pull request merged**